harness/determined/pytorch/_pytorch_context.py

```diff
@@ -73,6 +76,62 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
 
         self.experimental = pytorch.PyTorchExperimentalContext(self)
 
+    def autocast_forward_pass(self, to_wrap: torch.nn.Module) -> torch.nn.Module:
+        # First, ensure the forward pass is wrapped in an autocast context:
+        class _AutocastForwardPassModel(type(to_wrap)):  # type: ignore
+            def __init__(wrapper) -> None:
+                self.model = to_wrap
+
+            def __getattr__(wrapper, name):  # type: ignore
+                return getattr(to_wrap, name)
+
+            def __setattr__(wrapper, name, value):  # type: ignore
+                return setattr(to_wrap, name, value)
+
+            def __delattr__(wrapper, name):  # type: ignore
+                return delattr(to_wrap, name)
+
+            def forward(wrapper, *arg, **kwarg):  # type: ignore
+                with amp.autocast():
+                    return to_wrap.forward(*arg, **kwarg)
+
+        wrapped = _AutocastForwardPassModel()
+
+        # Second, eliminate any need for the loss functions to be in that context:
+        def end_fp16(
+            module: torch.nn.Module,
+            input: Any,
+            output: Any,
+            warned_types: Optional[Set[Type]] = None,
+        ) -> Any:
+
+            # Never print errors recursively.
+            if warned_types is None:
+                warned_types = set()
```

Comment:
This will prevent multiple calls per forward pass, but not multiple calls across forward passes.

I think if you put warned_types outside of `end_fp16` instead of having as a parameter this should work fine.

The reason `to_device()` does it differently is that `to_device` is defined at the library level but the warnings are tracked at the context level.  In this case, `end_fp` is defined once per model and that's probably a fine granularity to just use a single bound variable.